### ISSUE TITLE
Keystrokes involving ctrl on Linux resolve to the default layout instead of the active layout

### ISSUE THREAD

**Issue by allentiak** [atom/atom#13170]
### Prerequisites

Atom 1.12beta7 on Debian Stretch (testing).

### Description

My keyboard is using Colemak. Shortcuts only work in their QWERTY positions.

### Steps to Reproduce

1. Use Colemak layout
2. Open a file
3. Modify it
4. Try to save it via Colemak's Ctrl+S (which corresponds to the QWERTY position of Ctrl+D).

**Expected behavior:**
It should save the file (or prompt to save it as...)

**Actual behavior:**

It does not offer the options stated above.

** Workaround:**

Try to save it via QWERTY's Ctrl+S (which corresponds to the Colemak's position of Ctrl+R). It will work.

### Versions

Atom 1.12beta7 under Debian Stretch (testing).

**Comment by stevenhwu** [comment]
Not exactly relate to this, but I just want to point out that the shortcut key work as expected on Dvorak keyboard layout

My setup
Fedora 24 4.7.9-200.fc24.x86_64
atom 1.14.0-dev-c7f9a76

When I try to save via Dvorak's Ctrl+S, (Ctrl+: on QWERTY layout/standard keybord), it save the file as expected.

**Comment by nathansobo** [comment]
Can you post the results of typing some failing keystrokes after running the following devtools console script:

```
document.addEventListener('keydown', e => console.log(e), true)
```

Does it look like the `.key` field on the keyboard event is matching the keys you are actually pressing?

**Comment by allentiak** [comment]
It seems that the problem is when the shortcut involves Ctrl.

In the following code, Ctrl+Shift+P gets recognized as QWERTY's Ctrl+Shift+R.
```
KeyboardEvent {isTrusted: true, key: "Control", code: "ControlLeft", location: 1, ctrlKey: true…}
VM1197:1 
KeyboardEvent {isTrusted: true, key: "Shift", code: "ShiftLeft", location: 1, ctrlKey: true…}
VM1197:1 
KeyboardEvent {isTrusted: true, key: "R", code: "KeyR", location: 0, ctrlKey: true…}
```

However, if I only use Alt+F (Alt+E in QWERTY), it correctly shows the menu...
```
KeyboardEvent {isTrusted: true, key: "Alt", code: "AltLeft", location: 1, ctrlKey: false…}
VM1197:1 
KeyboardEvent {isTrusted: true, key: "f", code: "KeyE", location: 0, ctrlKey: false…}
```

**Comment by allentiak** [comment]
Update: I have just tested versions 1.12 and 1.13beta0, and the problem is still present in both.

**Comment by Gert-dev** [comment]
I've just encountered an additional, related, issue with Colemak after updating to 1.12: I can no longer use `Ctrl + Caps lock` as `Ctrl + Backspace` (the caps lock key corresponds to an additional backspace key on Colemak).

`Ctrl + Backspace` itself is working correctly and so is `Caps lock` itself (which removes a single character in front of the cursor position). In short only `Ctrl + Caps` lock is refusing to perform any action.

**Comment by nathansobo** [comment]
Wow. It's looking like there's a really bad bug in either Chrome or your Linux distribution that's causing incorrect key codes to be reported on Colemak when the ctrl key is used. I'm not sure how we can detect this bug in a general way since the keyboard events are just wrong.

This is unfortunately going to take a bit to resolve. Would anyone be interested in helping out in the meantime by using the [keystroke resolver](https://github.com/atom/atom-keymap/pull/164) API to build an ad-hoc workaround that Linux Colemak users could enable?

**Comment by nathansobo** [comment]
Could you also test this in the latest Chrome and report a bug to the Chromium project if you see a similar issue?

**Comment by Gert-dev** [comment]
I did a quick test with Chromium 54.0.2840.100 to see if it exposes the same issue. However, the `Ctrl + caps lock` combination seems to correctly resolve to a "delete word" action as expected here. Perhaps this issue has been solved in the meantime in Chromium as Atom 1.12 is currently at Chromium 52 with Electron 1.3.6?

I'm running on the official Atom Linux package at the moment (i.e. not from my distribution's repositories, but downloaded from the site), but I will also give the Atom from the official repositories a whirl once it updates to 1.12 as the official repositories seem to bundle the most recent electron version instead of the one expected by Atom. Perhaps electron 1.4 (Chromium 53) will fix the issue.

**Comment by nathansobo** [comment]
I will review and accept a PR that does an ad-hoc mapping of `KeyboardEvent.code` values when the Colemak layout is detected on Linux to work around the problem.

**Comment by GothAck** [comment]
Do you have multiple layouts set up? I found that Terminology and Atom preferred keyboard shortcuts (possibly only ones containing ctrl) in the "default" / primary keyboard layout. At least when using gnome:  just going into the language preferences and reordering the layout list sufficed.
This is still a workaround, whatever is causing this behavior still needs fixed, but for now it's stopped me throwing my laptop out of a window.

**Comment by GothAck** [comment]
This could also explain why some users have reported a-okay, their primary layout is the alt layout...

**Comment by Gert-dev** [comment]
@GothAck That would make sense, I'm afraid Colemak is already my main/first layout. However, I'm not suffering from any other issues than the control + backspace key not working, so perhaps that is an entirely different issue altogether.

**Comment by Ben3eeE** [comment]
@nathansobo Piecing together the information from reports it seems like we resolve keystrokes involving  <kbd>Ctrl</kbd> to the default layout instead of the active layout and it seems to happen for other layouts like azerty and qwertz as well. We should investigate if this is the case. I updated the title of the issue to reflect this.

**Comment by Ben3eeE** [comment]
I was able to reproduce this when using gnome on Ubuntu 16.04 but not when using unity.

**Steps to reproduce:**

1. Add `Swedish` and `French (Azerty)` as keyboard layouts in the gnome `Region and Language` settings
1. Change your default layout to `Swedish` and your active layout to `Swedish`
1. Start Atom and type keys involving the <kbd>Ctrl</kbd> modifier. Expected and Actual: Keys resolve correctly.
1. Change your default layout to `French (Azerty)` in the gnome `Region and Language` settings
1. Type keys involving the <kbd>Ctrl</kbd> modifier. Expected: All keystrokes to resolve using the `Swedish` layout because that is the active layout. Actual: Keystrokes using the <kbd>Ctrl</kbd> modifier resolve to the `French (Azerty)` layout all others resolve using the `Swedish` layout.

I tested with `English (US)`, `English (Dvorak)`, `Swedish` and `French (Azerty)` and all have this problem. As you rearrange layouts in the `Region and Language` settings the keys involving <kbd>Ctrl</kbd> resolve differently.

**Versions**
```
linus@linus:~$ atom --version
Atom    : 1.12.6
Electron: 1.3.9
Chrome  : 52.0.2743.82
Node    : 6.5.0

linus@linus:~$ apm --version
apm  1.12.9
npm  3.10.5
node 4.4.5
python 2.7.12
git 2.7.4

linus@linus:~$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.1 LTS
Release:	16.04
Codename:	xenial
```

**Comment by mattkang** [comment]
I use the Colemak layout as my primary input method. While I have not had issues with Ctrl shortcuts, my caps lock as backspace key does NOT work. This has been broken since 1.12. I am using the Colemak layout as provided [here](https://forum.colemak.com/topic/1621-colemak-for-windows-with-capslock-to-backspace/).

**Comment by Ben3eeE** [comment]
@mattkang @Gert-dev Please open a new issue for the caps lock backspace key not working on colemak in the https://github.com/atom/atom-keymap repository. This issue is a different one.

**Comment by mattkang** [comment]
@Ben3eeE atom/atom-keymap#192

**Comment by DavidCarpus** [comment]
@Ben3eeE  Any chance this is related to the mods from [atom/atom-keymap#51](https://github.com/atom/atom-keymap/pull/51)?
My dvorak has the same problem.
Any use of the ctrl key in a shortcut 'reverts' to standard qwerty but shortcuts that do not use ctrl seem to be ok.

**Comment by Ben3eeE** [comment]
This is unrelated to that PR. Are you on Linux? Gnome?

**Comment by DavidCarpus** [comment]
Ubuntu 16.01. Basic install. (All defaults in install) Added a few utils but no major mods to UI/X.

Sequences in other programs (Including Chrome browser Version 54.0.2840.100 (64-bit)) work.
I.E. Ctrl-T opens a new tab.

**Comment by Ben3eeE** [comment]
Yeah sounds like it might be this issue that you are facing. There is a fix for this in https://github.com/atom/atom/pull/13353 that we are considering hotfixing this on stable and beta. If people having this issue want to test this out the fix on master that would be great.

**Comment by Ben3eeE** [comment]
This has been fixed in Atom 1.12.7 available now ✨

**Comment by allentiak** [comment]
Thanks!

**Comment by toniopelo** [comment]
For my part, I have exactly the same issue even after updating to the latest release of atom.
```
$> atom -v
Atom    : 1.12.7
Electron: 1.4.12
Chrome  : 53.0.2785.143
Node    : 6.5.0
```
My keyboard layout is `French (Azerty)` and CTRL keys binds to `English (QWERTY)`

Antergos Linux (Arch linux based OS)
Gnome

**Comment by allentiak** [comment]
Which one of them is your default layout?

**Comment by toniopelo** [comment]
My OS is setup to work with the `French (AZERTY)` layout.
I don't know if this is what you mean by 'default layout'.

But for `CTRL + key` actions on atom it uses the English (QWERTY) layout.

**Comment by allentiak** [comment]
On 11 December 2016 at 11:15, toniopelo <notifications@github.com> wrote:
> My OS is setup to work with the French (AZERTY) layout.

Is this the only layout you have set up?

**Comment by toniopelo** [comment]
Yes, it seems so :
![keyboard-layout](https://cloud.githubusercontent.com/assets/9287511/21081040/cbcff280-bff8-11e6-8835-60c33361c51a.png)

Or maybe Gnome is playing with me but French is the only one in the list.
I can add others but they are not configured yet.

**Comment by pierrre** [comment]
Same issue.

I use:
- Atom 1.12.7
- Archlinux latest version
- Gnome
- French / AZERTY keyboard layout

I seems that "ctrl + xxx" shortcuts use the QWERTY layout instead of AZERTY.
"ctrl + z" => closes the current tab
"ctrl + w" => undo

**Comment by Ben3eeE** [comment]
>@Ben3eeE Maybe this issue should be reopened ?

You are using an [unoffical build of Atom](https://github.com/atom/design-decisions/blob/master/unofficial-atom-distributions.md) which uses a mismatched version of Electron. This issue got fixed by upgrading Electron to a version that included the fix from chromium so that might be why you are having this issue.

Can you paste the following script into the developer tools console and post the results of pressing the failing keystrokes:
```
document.addEventListener('keydown', e => console.log(e), true)
```

**Comment by toniopelo** [comment]
Oh..
I wasn't aware I was using an unofficial build.. I installed it from the community repo of archlinux.

There are the two console outputs when pressing `CTRL+Z` that finally triggers `CTRL+W`:
![console1](https://cloud.githubusercontent.com/assets/9287511/21081554/7dc150a0-c004-11e6-9c25-aac9c5d8b9ea.png)
![console2](https://cloud.githubusercontent.com/assets/9287511/21081555/7dc502e0-c004-11e6-9e5a-540be67f039b.png)

**Comment by 50Wliu** [comment]
> I installed it from the community repo of archlinux.

Yeah, many of the Arch Linux community repos refuse to use the correct version of Electron, leading to these kinds of issues.  I highly recommend that you uninstall Atom and reinstall it from https://atom.io.

**Comment by toniopelo** [comment]
Yeah sure!
But I will loose the ability to update it automatically right ?
I will need to download it again and again for every new release ?

**Comment by MonsieurLanza** [comment]
Same problem here, same fix. 
Thanks @toniopelo.

**Comment by Ben3eeE** [comment]
@toniopelo Looking at the event it seems the version of Atom you were using before does not include the fix that closed this issue. If you want to continue using the version from the community repo you should open an issue there for this bug.

If the `atom-editor-bin` works for you then it is a good solution :slightly_smiling_face:. Just to confirm does `atom --version` of this say it's using `Electron 1.3.13`?

**Comment by MonsieurLanza** [comment]
As I just installed `atom-editor-bin` from AUR myself :
```bash
$ atom --version
Atom    : 1.12.7
Electron: 1.3.13
Chrome  : 52.0.2743.82
Node    : 6.5.0
```
The "regular" community `atom` package showed exact same version as @toniopelo.

**Comment by toniopelo** [comment]
There it is :
```
$ atom --version
Atom    : 1.12.7
Electron: 1.3.13
Chrome  : 52.0.2743.82
Node    : 6.5.0
```

So I'm using an older electron version than before right ?
On my first post I was apparently using electron 1.4.12

**Comment by Ben3eeE** [comment]
Yeah, the package in the arch community repository uses a newer version of Electron and other built in components that can cause various issues. The current Electron version for Atom is 1.3.13.

Happy to hear it is working for you when using the AUR package :slightly_smiling_face:

**Comment by toniopelo** [comment]
Good :)
Thank you for all!

**Comment by pierrre** [comment]
FYI https://bugs.archlinux.org/task/52120

**Comment by Ben3eeE** [comment]
@nathansobo Can we close this now that https://github.com/atom/atom/pull/13543 has shipped?

**Comment by nathansobo** [comment]
Closed via #13543

**Comment by MrGreenPepper** [comment]
moving  the new layout in the primary position fixed the problem for me
